core/kernel/delay.c

```diff
@@ -32,6 +32,20 @@
 #include <kernel/misc.h>
 
 #ifdef CFG_CORE_HAS_GENERIC_TIMER
+int timeout_elapsed_us(uint64_t expire)
+{
+	int64_t diff = delay_cnt_read() - expire;
+
+	if (MUL_OVERFLOW(diff, ULL(1000000), &diff)) {
+		if (timeout_elapsed(expire))
+			return INT_MAX;
+		else
+			return INT_MIN;
+	}
+
+	return (int)(diff / (uint64_t)delay_cnt_freq());
```

Comment:
Why is any casting needed? Casting the result from `delay_cnt_freq()` to the same type as `diff` might make sense, but I'm not sure that there's a point in even that.

Reply:
I guess casting `delay_cnt_freq()` is not needed. As for casting the result to an `int`, maybe I should trap them in the above test.